x/emissions/keeper/msgserver/msg_server_stake_test.go

```diff
@@ -817,3 +817,223 @@ func (s *KeeperTestSuite) TestRewardDelegateStake() {
+	require.NoError(err)
+	require.NotNil(responseDelegator, "Response should not be nil after successful delegation")
+
+	var reputerValueBundles types.ReputerValueBundles
+	scoreToAdd := types.Score{
+		TopicId:     topicId,
+		BlockNumber: block,
+		Address:     reputerAddr.String(),
+		Score:       score,
+	}
+	err = keeper.InsertReputerScore(s.ctx, topicId, block, scoreToAdd)
+	s.Require().NoError(err)
+
+	reputerValueBundle := &types.ReputerValueBundle{
+		ValueBundle: &types.ValueBundle{
+			TopicId:       topicId,
+			Reputer:       reputerAddr.String(),
+			CombinedValue: alloraMath.MustNewDecFromString("1500.0"),
+			NaiveValue:    alloraMath.MustNewDecFromString("1500.0"),
+		},
+	}
+	reputerValueBundles.ReputerValueBundles = append(reputerValueBundles.ReputerValueBundles, reputerValueBundle)
+	err = keeper.InsertReputerLossBundlesAtBlock(s.ctx, topicId, block, reputerValueBundles)
+	s.Require().NoError(err)
+
+	// Calculate and Set the reputer scores
+	scores, err := rewards.GenerateReputerScores(s.ctx, s.emissionsKeeper, topicId, block, reputerValueBundles)
+	s.Require().NoError(err)
+
+	// Generate rewards
+	reputers, reputersRewardFractions, err := rewards.GetReputersRewardFractions(
+		s.ctx,
+		keeper,
+		topicId,
+		alloraMath.OneDec(),
+		scores,
+	)
+	s.Require().NoError(err)
+	reputerRewards, err := rewards.GetRewardPerReputer(
+		s.ctx,
+		keeper,
+		topicId,
+		alloraMath.MustNewDecFromString("1017.5559072418691"),
+		reputers,
+		reputersRewardFractions,
+	)
+	s.Require().NoError(err)
+	s.Require().Equal(1, len(reputerRewards))
+
+	delegatorBal0 := s.bankKeeper.GetBalance(ctx, delegatorAddr, params.DefaultBondDenom)
+	rewardMsg := &types.MsgRewardDelegateStake{
+		Sender:  delegatorAddr.String(),
+		TopicId: topicId,
+		Reputer: reputerAddr.String(),
+	}
+	_, err = s.msgServer.RewardDelegateStake(ctx, rewardMsg)
+
+	delegatorBal1 := s.bankKeeper.GetBalance(ctx, delegatorAddr, params.DefaultBondDenom)
+	s.Require().NoError(err)
+
+	s.Require().Greater(delegatorBal1.Amount.Uint64(), delegatorBal0.Amount.Uint64(), "Balance must be increased")
+
+	delegatorReward := delegatorBal1.Amount.Sub(delegatorBal0.Amount)
+	reputerReward := reputerRewards[0].Reward.SdkIntTrim()
+
+	s.Require().Equal(delegatorReward, reputerReward, "Delegator and reputer rewards must be equal")
+}
+
+func (s *KeeperTestSuite) Test1000xDelegatorStakeVsReputerStake() {
+	ctx := s.ctx
+	require := s.Require()
+	keeper := s.emissionsKeeper
+	block := int64(1003)
+	// newBlock := int64(1004)
+	score := alloraMath.MustNewDecFromString("17.53436")
+
+	delegatorAddr := sdk.AccAddress(PKS[0].Address())
+	reputerAddr := sdk.AccAddress(PKS[1].Address())
+	workerAddr := sdk.AccAddress(PKS[2].Address()) // target
+
+	registrationInitialBalance := cosmosMath.NewInt(10000)
+	reputerStakeAmount := cosmosMath.NewInt(1e2)
+	delegatorRatio := cosmosMath.NewInt(1e3)
+	delegatorStakeAmount := reputerStakeAmount.Mul(delegatorRatio)
```

Comment:
I think this test is fine but I'd add in another that does something more like the following:
1. Have a reputer and delegator
2. Have them earn some rewards for 1-2 rounds.
   a. Record what each is entitled to receive as of that moment.
3. Have a 1000x delegator enter
4. Do another reward round
5. See if that 1000x delegator can withdraw more than they should, "stealing rewards" because they rebalance the current proportion of delegator stake to be in their favor
   a. So, the 1000x delegator should receive just rewards from the round they were in, NOT that PLUS any rewards that were otherwise payable to existing reputer or delegator

I'd add in 1 additional delegator to this check that delegates less / comparable to the reputer, as there is some accounting that is unique to how their rewards are counted that may admit this attack.

ik just a draft, just letting you know my thoughts in case helpful

Reply:
definitely helpful, thanks! Im at the stage of figuring out which more complex tests to run, Ill add this one.